### PR TITLE
Add release workflow

### LINKED ISSUE #730: Trigger releases via GitHub Actions
Let's see if we can release EqualsVerifier to Maven Central through a button on GitHub Actions

### PR DESCRIPTION
**Pull request by scordio**
Fixes #730.

Once merged, the release can be triggered via [manual action](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow) under the [workflow page](https://github.com/jqno/equalsverifier/actions/workflows/release.yml).

The version should be set before triggering the release as no special parameter is exposed (the released version will be the one set in the POM and the next development version will have the patch number increased).

The following secrets should be added to the [repository configuration](https://github.com/jqno/equalsverifier/settings/secrets/actions):

* `GPG_PASSPHRASE`
* `GPG_PRIVATE_KEY`
* `OSSRH_TOKEN`
* `OSSRH_USERNAME`

For general transparency about [GitHub secret management](https://docs.github.com/en/actions/security-guides/encrypted-secrets) and related security concerns, GitHub uses a [libsodium sealed box](https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes) to help ensure that secrets are encrypted before they reach GitHub and remain encrypted until you use them in a workflow.

### CODE REVIEW

**Comment by scordio** [comment]
From https://github.com/jqno/equalsverifier/issues/730#issuecomment-1319801851:

> Assume all versions are set properly in the POM file (or maybe supplied via a GUI?) - in that case ideally the README.md and docs/_includes/maven files should also be updted. Maybe something for later :).

I suggest not having the version value as a workflow input, although it could be done. I consider using the current value in the POM less error-prone. As mentioned in the description, the version should be set before triggering the release as no special parameter is exposed (the released version will be the one set in the POM and the next development version will have the patch number increased). This means that README and docs should be manually updated right before releasing.

> Make sure we're running Java 17

Specified in the workflow ✅ 

> Make sure all builds are green

As the workflow is manually triggered, I'd suggest this will be a manual check.

> `git checkout -b release-<version number>`
> `mvn -s ~/.m2/equalsverifier-deploy-settings.xml release:prepare release:perform` (the file contains credentials for Sonatype Nexus and a GPG password)

The deploy settings will be configured by the `setup-java` action (more [here](https://github.com/actions/setup-java/blob/main/docs/advanced-usage.md#Publishing-using-Apache-Maven)) and GitHub secrets should contain the necessary data.

However, I don't see any declaration of:
* `maven-gpg-plugin`, usually involved in signing
* `nexus-staging-maven-plugin`, usually involved in deployment to the Sonatype Nexus

Do you have any of these in the local execution logs?

> `git push && git push --tags`
> `git checkout main`
> `git merge -no-ff release-<version number>`
> `git push`
> `git push --tags`

All these I believe should be done by the [maven-release-plugin](url) automatically, that's why I removed `pushChanges` and `localCheckout` parameters to make sure we stick to the default.

**Comment by scordio** [comment]
@jqno I've commented on all your points, happy to get your feedback.

**Comment by jqno** [comment]
Wow, that was fast! Thank you!

I agree with the manual update of the poms and the docs. I also agree with manually checking that the builds are green. It's how I've always done it, and it's no big deal anyway. I also agree with removing `pushChanges` and `localCheckout` from the pom. When I initially set up publishing to Maven Central over 10 years ago (:scream:), I had some problems if I didn't do it, but I don't remember what they were. Probably I was still anxious about scripting this, and I wanted to have some manual control.

Regarding `maven-gpg-plugin` and `nexus-staging-maven-plugin`, the answer is that I don't know. Unfortunately I closed the window I used this morning to make a release, so I don't have the logs anymore :grimacing: I think the `maven-release-plugin` calls `maven-gpg-plugin` automagically? I think `maven-deploy-plugin`, which is enabled by default and not visible in the pom, will take care of uploading everything to Maven Central, so I never set anything up for that, specifically. Again, this was all 10 years ago, workflows have probably improved since then :sweat_smile: . When it's uploaded, I have to manually log in to Sonatype Nexus and push a few buttons before it's actually released.

I guess I also have to figure out how to get an `OSSRH_TOKEN` and `OSSRH_USERNAME` :thinking:

**Comment by scordio** [comment]
> Regarding `maven-gpg-plugin` and `nexus-staging-maven-plugin`, the answer is that I don't know. Unfortunately I closed the window I used this morning to make a release, so I don't have the logs anymore 😬 I think the `maven-release-plugin` calls `maven-gpg-plugin` automagically? I think `maven-deploy-plugin`, which is enabled by default and not visible in the pom, will take care of uploading everything to Maven Central, so I never set anything up for that, specifically. Again, this was all 10 years ago, workflows have probably improved since then 😅 . When it's uploaded, I have to manually log in to Sonatype Nexus and push a few buttons before it's actually released.

I'm no expert here, I've just stolen configurations over time 😅 I can offer to replicate the config we have in AssertJ which worked fairly well so far, the same as what japicmp (siom79/japicmp@54f0c464a9db1e76de6086fc1da0a0b1542abb83) and ByteBuddy do (ByteBuddy has been my inspiration when I've started with this topic in AssertJ).

> I guess I also have to figure out how to get an OSSRH_TOKEN and OSSRH_USERNAME

1. https://oss.sonatype.org/
2. Log In
3. Profile
4. Select "User token" from the "Summary" dropdown
5. Access User Token

You'll get a window with the username and the token.

**Comment by jqno** [comment]
I also often use ByteBuddy as a template for these kinds of things :smile: 

I've added the secrets to the repository. I won't be at my laptop much this weekend, I think any next steps on my part will probably have to happen on Monday. I am very excited by this though! Thanks for all your help! :smile:

**Comment by scordio** [comment]
No problem, as promised I'll enhance this PR with the remaining config and switch it as ready for review afterward.

**Comment by scordio** [comment]
> When it's uploaded, I have to manually log in to Sonatype Nexus and push a few buttons before it's actually released.

Probably that's the role of the `nexus-staging-maven-plugin`, again, no expert here, just guessing 😆

**Comment by jqno** [comment]
Ahh that actually makes a lot of sense, I'll have to look into that 😄

**Comment by scordio** [comment]
I realized only now that you're using the `oss-parent` version 9 from Sonatype so this at least explains how the `maven-gpg-plugin` gets executed.

However, the [GitHub homepage](https://github.com/sonatype/oss-parents) states:

> This project is no longer active or supported. We suggest to manage parent POM files for your own organization as needed. The POM files from this project no longer work with latest Maven and/or Java versions.

so I guess it's a good motivation to renovate these aspects 😄

**Comment by jqno** [comment]
Haha oops, didn't know it was deprecated :sweat_smile: 
I should probably remove it then, and see if I can still deploy...

Looks like AssertJ indeed uses `nexus-staging-maven-plugin` for releasing to Maven Central: https://github.com/assertj/assertj/blob/main/pom.xml#L235-L244

In particular the `<autoReleaseAfterClose>` line is interesting: I recognise it as one of the manual steps I have to perform on Maven Central. I'll have to look into this a bit closer after the weekend.

**Comment by scordio** [comment]
> Looks like AssertJ indeed uses `nexus-staging-maven-plugin` for releasing to Maven Central: https://github.com/assertj/assertj/blob/main/pom.xml#L235-L244

Of course AssertJ uses it, guess who did [the latest changes](https://github.com/assertj/assertj/commit/c6543878b3c606e39093c494dcd62178e47c2d79) 😄 

I'll adjust this PR to be a bit more than just the addition of a new workflow!

File: pom.xml [hunk 1]
```diff
@@ -194,7 +211,7 @@
                 <version>${version.maven-source-plugin}</version>
                 <executions>
                     <execution>
-                        <phase>package</phase>
```
Comment (scordio) [inline pom.xml:197]: Removed as `jar-no-fork` binds by default to the `package` phase.

**Comment by scordio** [comment]
I hope I haven't missed anything, this is pretty much mimicking what we have in AssertJ.

**Comment by scordio** [comment]
Minor point: the [`tidy-maven-plugin`](https://www.mojohaus.org/tidy-maven-plugin/index.html) can help ordering the POM according to the [Maven conventions](https://maven.apache.org/developers/conventions/code.html#pom-code-convention).

**Comment by jqno** [comment]
> Of course AssertJ uses it, guess who did [the latest changes](https://github.com/assertj/assertj/commit/c6543878b3c606e39093c494dcd62178e47c2d79) :smile: 

Haha, of course, I should have realised :).

> I hope I haven't missed anything, this is pretty much mimicking what we have in AssertJ.

The only way to know, is to merge and try it out, I suppose :smiley: 
I suppose, with this workflow, I still have to log in to oss.sonatype.org to actually close the release, right? If so, I can safely run the pipeline without actually releasing something. I could do that today or tomorrow.
I have some time reserved for work on EqualsVerifier next week, so most probably I can make a real release then using the new pipeline.

> Minor point: the [tidy-maven-plugin](https://www.mojohaus.org/tidy-maven-plugin/index.html) can help ordering the POM according to the [Maven conventions](https://maven.apache.org/developers/conventions/code.html#pom-code-convention).

Does that apply to the pom files in my project, or the pom file that will be published by the release? Either way, looks interesting, I didn't know it exists!

**Comment by Stephan202** [comment]
> Does that apply to the pom files in my project, or the pom file that will be published by the release

The ones in the project: there's a goal to check and a goal to fix. :)

**Comment by scordio** [comment]
> I suppose, with this workflow, I still have to log in to oss.sonatype.org to actually close the release, right? If so, I can safely run the pipeline without actually releasing something. I could do that today or tomorrow.

No, with this pipeline no manual action is required. In AssertJ, we just trigger the pipeline manually and the release is done end-to-end.

About the `tidy-maven-plugin`, I can add the needed config but maybe in a separate PR!

**Comment by jqno** [comment]
OK, so in that case, maybe I should just merge and wait with testing until I actually have something to release, which should (hopefully) be some time next week.

>About the tidy-maven-plugin, I can add the needed config but maybe in a separate PR!

Well, I appreciate that, but I kinda feel bad for asking, you've been doing so much already!

**Comment by scordio** [comment]
No worries, it's no big deal :-)

**Comment by jqno** [comment]
Well, this PR looks good to me. I'm just going to merge, and I'll pick up something small to work on next, so I can test the release quickly. :smile:

**Comment by jqno** [comment]
Thank you so much for all this work, I really appreciate it!

**Comment by scordio** [comment]
> No worries, it's no big deal :-)

Of course, I was too optimistic 😅

I tried to enforce the POM formatting by adding an execution of the [`check`](https://www.mojohaus.org/tidy-maven-plugin/check-mojo.html) goal of the `tidy-maven-plugin` directly in the root POM and any inconsistencies could be fixed by manually executing the [`pom`](https://www.mojohaus.org/tidy-maven-plugin/pom-mojo.html) goal.

However, the `maven-shade-plugin` produces a `dependency-reduced-pom` which is picked up by the verification of the `tidy-maven-plugin` and of course doesn't follow the formatting rules the plugin expects.

I don't see a way to make it work unless I introduce automated formatting of the generated POM, right after the `maven-shade-plugin` execution. I'm not a big fan of automated modifications so I suggest dropping the idea.

**TL;DR** while it could be nice to have the POM sections ordered and automatically verified, I don't see an easy to enforce this because of the `maven-shade-plugin`.

**Comment by jqno** [comment]
Aw, that's unfortunate, but I agree with your suggestion. Maybe when Maven 4 (5?) is out and we can start having a cleaner separation between project pom and published pom, we can reconsider.

Still, thank you so much for trying, and all the other work you did! I'll let you know how the release goes text time!

**Comment by jqno** [comment]
I'm sad to report that I had to revert to the old, manual way to release EqualsVerifier just now :disappointed: 

I ran into this bug in the nexus-staging-maven-plugin: https://issues.sonatype.org/browse/NEXUS-9138

The problem is that EqualsVerifier has a multi-module build, and only two submodules should ever be deployed. You can influence this using the ` skipNexusStagingDeployMojo` configuration setting, but for some reason, it will deploy NOTHING if the last module isn't also deployed. In the case of EqualsVerifier, that's the `equalsverifier-release-verify` module, which runs some checks against the prepared artifacts and produces nothing that needs to be deployed.

I tried a bunch of things to work around this, including removing the `equalsverifier-release-verify` module from the build entirely if a release profile is active, but even then it gave me errors. For now I've parked your commits in the `automatic-release` branch, hopefully I can give it another try soon.

**Comment by scordio** [comment]
So sad to hear that, also because I've recently switched AssertJ to a multi-module config and we might face the same issues in the next release as not all the artifacts should be deployed 😢 

At least I know that I need to reserve some more time for the next release and dig a bit more, I'll let you know if I find something that can help also here.

**Comment by jqno** [comment]
I was thinking of looking into JReleaser instead of nexus-staging-maven-plugin, which is now able to deploy to Maven Central as well. No idea if it even supports skipping specific submodules, but at least the developer is very responsive :). And also, JReleaser has some other capabilities as well that could be useful, like automatic tweeting/tooting after a release.

**Comment by jqno** [comment]
I hope you can make it work for AssertJ. Unfortunately I don't have a lot of time to look into this right now myself, otherwise I would, and let you know what else to look out for. Sorry...

**Comment by scordio** [comment]
> I was thinking of looking into JReleaser instead of nexus-staging-maven-plugin, which is now able to deploy to Maven Central as well. No idea if it even supports skipping specific submodules, but at least the developer is very responsive :). 

Yes, @aalmiray is very supportive and the JReleaser documentation is thorough. This could really be the right excuse to switch to JReleaser 🚀

**Comment by aalmiray** [comment]
The `nexus-staging-plugin` works in this way when in a multi-module build: changes are committed to the staging repository when the last module is executed. It looks like for EqualsVerifier the last module is skipped hence why no artifacts get published to the staged repository.

In the case of JReleaser there have been a couple of issues fixed since `1.3.0` (currently targeting `1.4.0` at the end of Dec 2022) that close the gap with Nexus2 support (required to publish to Maven Central). Given that EqualsVerifier and AssertJ will likely use the `jreleaser-maven-plugin` I'd suggest you to wait til `1.4.0` comes out. Alternatively, you may build your own snapshot of `jreleaser-maven-plugin` and run some local tests.

FWIW this is what needed to publish artifacts to Maven Central using JReleaser https://jreleaser.org/guide/latest/examples/maven/index.html

**Comment by aalmiray** [comment]
More over, 2 additional examples of real-life usage:

- https://github.com/moditect/layrry
- https://github.com/kordamp/pomchecker

Both projects publish only a subset of all Maven modules, which should match characteristics expected by EqualsVerifier.

**Comment by jqno** [comment]
Awesome, thanks for that @aalmiray ! I think I'll wait for 1.4.0, as you suggest.
BTW, is there a way to test the integration without actually releasing something?

**Comment by aalmiray** [comment]
Yes, you can run JReleaser in dry-run mode, which will skip remote operations on write (push). For the Maven plugin it's enough to set `jreleaser.dry.run` tor `true` either as a POM property or in the command line with `-Djreleaser.dry.run=true`.

**Comment by jqno** [comment]
Thanks for that, that looks good!
This seems like a great hobby project for the christmas holidays :smile:

**Comment by aalmiray** [comment]
One of the goals of JReleaser is to enable you to experiment with releases on a local environment, hence the option to run it in dry-run mode. You don't have to run it on CI all the time to test out a theory 😏

**Comment by jqno** [comment]
I've just made my first official release using JReleaser! :rocket: I hope everything went well :crossed_fingers: :sweat_smile: 

@scordio If you want to move AssertJ to use JReleaser as well, here's the guide I followed: https://foojay.io/today/how-to-release-a-java-module-with-jreleaser-to-maven-central-with-github-actions/
It's actually pretty straightforward, but there are some things to look out for:
- The author assumes a new project; you probably have to use `https://oss.sonatype.org/service/local` instead of `https://s01.oss.sonatype.org/service/local` in the configuration. According to @aalmiray's [excellent docs](https://jreleaser.org/guide/latest/examples/maven/maven-central.html#_maven), this was changed in early 2021.
- In the GitHub workflow section, the author has a bunch of workflow snippets, and then refers to his final version of the file. Beware that if you copy paste that, it contains some changes compared to what's on the website, most notably the name of the `JRELEASER_NEXUS2_MAVEN_CENTRAL_USERNAME` variable. That tripped me up!
- Also, in one of the last snippets, he explicitly pushes the tag, but JReleaser already does this for you, so you'll get an error at that point. (Also, in the full file, he adds a manual step to create the tag, but of course JReleaser does that as well :).)
- I had to add `jreleaser-cli.jar` to `.gitignore`. The last step in the github workflow committed it for some reason and I had to force push to get it out again. Don't want a large binary cluttering up your repo...

Anyway, if you have questions, maybe this time I can help you instead of the other way round :smile: 

I still have some homework for myself:

- Backport the changes you made to the pom; a lot of that is still valuable even if I don't need it for JReleaser.
- Explore the JReleaser docs some more to see if I can improve or streamline the process. So many options, so many features! The generated changelog is a bit messy so I should look into improving that, and maybe I can have it publish something on Mastodon and Twitter for me.

**Comment by jqno** [comment]
Also, @aalmiray , if you're still reading this :sweat_smile: 
I got some warnings of the form `Cannot parse version 'ersion_0_1'` while releasing. I used to use tags of the form `version_0_1` early in the project; later I switched to the `equalsverifier-3.12.4` format (which I've successfully encoded in the jreleaser config). Could it be that these tags cause problems for JReleaser somehow?
Here's the relevant section of the log: https://github.com/jqno/equalsverifier/actions/runs/4033565937/jobs/6934172332#step:7:82

**Comment by jqno** [comment]
I have to say, all in all, even without the ability to release from GitHub pages, I'm quite happy with JReleaser. Deploying with Maven's release plugin was always so twiddly, this is a lot simpler and feels more robust somehow.

**Comment by aalmiray** [comment]
> Also, @aalmiray , if you're still reading this 😅 I got some warnings of the form `Cannot parse version 'ersion_0_1'` while releasing. I used to use tags of the form `version_0_1` early in the project; later I switched to the `equalsverifier-3.12.4` format (which I've successfully encoded in the jreleaser config). Could it be that these tags cause problems for JReleaser somehow? Here's the relevant section of the log: https://github.com/jqno/equalsverifier/actions/runs/4033565937/jobs/6934172332#step:7:82

These warnings appear because old tags do not conform to the new format. I'm afraid they'll continue to appear on each release. One way to solve this is to rename the tags which can be done by following the instructions found at https://stackoverflow.com/a/5719854/120745.

**Comment by jqno** [comment]
Wow, that was quick. Thank you! I'll see if it bothers me enough to actually do that, but at least now I know how to fix it. Thank you :)

**Comment by scordio** [comment]
Great job, congrats to both of you @jqno and @aalmiray! 🙂 

In the latest version of AssertJ, I obviously faced the same problem but I procrastinated switching to JReleaser with assertj/assertj@b240643c1ffa63be46fa1b2ea9f6ae326810a35e, ugly as hell but allowed me to release without major problems in the reduced window I had available.

As soon as I get some time, we will definitely make the switch.

**Comment by jqno** [comment]
That's not pretty, but worming code is more important than beautiful code...
Anyway, let me know if I can be of any assistance. Happy to help!
